### PR TITLE
Validate Subcommand!

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR introduces to bom the capability to start validating things. For now, only files attached to the first level of the document are supported and we will be introducing more artifact types and other kinds of validation.

image

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

~~Test for `document.ValidateFiles()` is coming in a followup shortly~~

Test is now included

/assign @cpanato 

#### Does this PR introduce a user-facing change?

```release-note
`bom` can now validate artifacts! We now have a new validate subcommand that can be used to check files attached to the top of the SBOM: `bom validate sbom.spdx file.txt`. No more checksum.txt files! 🎉
```

### CODE REVIEW

File: cmd/bom/cmd/validate.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+type validateOptions struct {
+	exitCode bool
+	sbomPath string
+	files    []string
+}
+
+// Validate verify options consistency
+func (opts *validateOptions) Validate() error {
+	if len(opts.files) == 0 {
+		return errors.New("plaease provide at least one artifact to validate")
```
Comment (cpanato) [inline cmd/bom/cmd/validate.go:112]:
```suggestion
		return errors.New("please provide at least one artifact to validate")
```

**Comment by cpanato** [comment]
thanks for this feature

File: cmd/bom/cmd/root.go [hunk 2]
```diff
@@ -62,6 +62,7 @@ func init() {
 	rootCmd.AddCommand(generateCmd)
 	rootCmd.AddCommand(documentCmd)
 	rootCmd.AddCommand(versionCmd)
+	AddValidate(rootCmd)
```
Comment (jeremyrickard) [inline cmd/bom/cmd/root.go:65]: I noticed that this deviates from the pattern of the other commands. Do you plan on refactoring the others to follow this new pattern?

Comment (puerco) [reply cmd/bom/cmd/root.go:65]: Yep, tried to switch to the pattern @justaugustus recommended here: https://github.com/kubernetes-sigs/bom/pull/27#pullrequestreview-860236487
I will refactor the rest of the CLI to match.

Comment (jeremyrickard) [reply cmd/bom/cmd/root.go:65]: 👍 nice

**Comment by puerco** [comment]
Ready,foxed the nit and also pushed the new function test. Please take another look when you have time @cpanato

File: cmd/bom/cmd/validate.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		"list of files to verify",
+	)
+
+	cmd.PersistentFlags().BoolVarP(
+		&valOpts.exitCode,
+		"exit-code",
+		"e",
+		false,
+		"when true, bom will exit with exit code 1 if invalid artifacts are found",
+	)
```
Comment (justaugustus) [inline cmd/bom/cmd/validate.go:98]: I think this is maybe extraneous?
Is there a scenario where we'd want a `validate` command to "fail successfully"?

**Review by justaugustus — CHANGES_REQUESTED** [review]
@puerco -- One nit/thing that needs clarification!

File: cmd/bom/cmd/validate.go (hunk 3 above)
Comment (puerco) [reply cmd/bom/cmd/validate.go:98]: Yes! The idea is that you check the integrity of files you download and use the tool in a script:

`bom validate -e sbom.spdx file.zip || echo "Your zipfile is corrupt"`  

The bom action will use it to verify bom itself:

https://github.com/puerco/bom-installer/blob/1d65808b8b512ca77cf2d73cb844570cfec9f39e/action.yaml#L161

**Comment by puerco** [comment]
Fixed the merge conflicts :rocket:

**Review by saschagrunert — APPROVED** [review]

**Pull request merged**